threeML/utils/spectrum/spectrum_likelihood.py

```diff
@@ -0,0 +1,315 @@
+
+        _, background_model_counts = self.get_log_likelihood()
+
+        randomized_background_counts = np.random.poisson(background_model_counts)
+
+        return randomized_background_counts
+
+class PoissonObservedGaussianBackgroundLikelihood(SpectrumLikelihood):
+
+    def get_log_likelihood(self):
+
+        expected_model_counts = self._spectrum_plugin.get_model()
+
+        loglike, bkg_model = poisson_observed_gaussian_background(self._spectrum_plugin.current_observed_counts,
+                                                                  self._spectrum_plugin.current_background_counts,
+                                                                  self._spectrum_plugin.current_background_count_errors,
+                                                                  expected_model_counts)
+
+        return np.sum(loglike), bkg_model
+
+    def get_randomized_source_counts(self, source_model_counts):
+        # Since we use a profile likelihood, the background model is conditional on the source model, so let's
+        # get it from the likelihood function
+
+        _, background_model_counts = self.get_log_likelihood()
+
+        # Now randomize the expectations
+
+        # Randomize expectations for the source
+
+        randomized_source_counts = np.random.poisson(source_model_counts + background_model_counts)
+
+        return randomized_source_counts
+
+
+    def get_randomized_background_counts(self):
+
+        # Now randomize the expectations.
+
+        _, background_model_counts = self.get_log_likelihood()
+
+        # We cannot generate variates with zero sigma. They variates from those channel will always be zero
+        # This is a limitation of this whole idea. However, remember that by construction an error of zero
+        # it is only allowed when the background counts are zero as well.
+        idx = (self._spectrum_plugin.background_count_errors > 0)
+
+        randomized_background_counts = np.zeros_like(background_model_counts)
+
+        randomized_background_counts[idx] = np.random.normal(loc=background_model_counts[idx],
+                                                             scale=self._spectrum_plugin.background_count_errors[idx])
+
+        # Issue a warning if the generated background is less than zero, and fix it by placing it at zero
+
+        idx = (randomized_background_counts < 0)  # type: np.ndarray
+
+        negative_background_n = np.sum(idx)
+
+        if negative_background_n > 0:
+            custom_warnings.warn("Generated background has negative counts "
+                                 "in %i channels. Fixing them to zero" % (negative_background_n))
+
+            randomized_background_counts[idx] = 0
+
+        return randomized_background_counts
+
+    def get_randomized_background_errors(self):
+
+        return copy.copy(self._spectrum_plugin.background_count_errors)
+
+
+likelihood_lookup = {'poisson':{'poisson' : PoissonObservedPoissonBackgroundLikelihood,
```

Comment:
If the user wants to add its own statistic, he/she can subclass the SpectrumLikelihood class. But then, how can he use the new statistic? I guess he/she will need to add it to this dictionary, but then how is he going to set it for the spectrum?

Reply:
I can think of a way to do this. Sort of like how you can add models in astromodels. However, I don't personally like the idea of custom likelihoods. This is the problem with XSPEC. I understand that we allow for the changing of the noise models (we should add a warning!), but allowing too much freedom and we get 3ML Chandra papers with bonkers statistics.

Reply:
I understand your point, but we should also allow the advanced users the freedom they deserve. Say that somebody needs a statistic based on the Skellam distribution... or maybe somebody wants to experiment with some custom analysis and they need some weird statistic...

When we added PGstat in Xspec, I had to personally write Keith Arnaud and ask him to implement it. He was very kind and did it in a very reasonable time frame, but I don't think this is optimal.

Reply:
I agree. It's open source, so there is always the chance. I think a good compromise would be to to retain the original data's statistical properties and warn the user pretty blatantly that the choice being made is not reflecting the data's distribution.